core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -322,15 +325,37 @@ class ControllerApis(val requestChannel: RequestChannel,
           }
         }
         // Finally, the idToName map contains all the topics that we are authorized to delete.
-        // Perform the deletion and create responses for each one.
-        controller.deleteTopics(context, idToName.keySet).thenApply { idToError =>
-          idToError.forEach { (id, error) =>
-            appendResponse(idToName.get(id), id, error)
+        // First check the controller mutation quota if necessary, and then
```

Comment:
This is a lot more logic (and expense) than what I wanted, and unfortunately it also has a TOCTOU, because partitions may be added in between checking here and doing the deletion.

I guess I simply didn't forsee this, sorry.

I wonder if we could convert the throttle to be lockless so that we could use it in the QuorumController thread itself. Like I said earlier, the big concern is that blocking the controller thread is really bad and not something we want to do under really any condition. It seems like the throttle should be simple... just a number or something right? Or can we somehow guarantee that only the controller thread itself is taking that lock under ordinary conditions?

Reply:
I believe these locks are scoped by the controller mutation quota manager such that nobody else would contest for them aside from the single-threaded controller.  @dajac might you be able to confirm this?

Reply:
The quota is basically a `Sensor` which would only be updated by the QuorumController thread. However, I suppose that the internal metric lock is also acquired when the metric is reported so the QuorumController would not be the only one trying to acquire it, I think.